include/zio/zio_defs.h

```diff
@@ -0,0 +1,510 @@
+	 * reference point.
+	 *
+	 * Expressed in meters (m).
+	 */
+	DEV_CHANNEL_POS_DIST_SURF,
+
+	/**
+	 * Proximity to an external plane or object.
+	 *
+	 * Expressed in millimeters (mm).
+	 */
+	DEV_CHANNEL_POS_PROX,
+
+	/**
+	 * Encapsulates a raw NMEA sentence from a GPS-type device
+	 *
+	 * Expressed as a null-terminated string, with the exact string
+	 * contents determined by NMEA standard. NMEA sentences are generally
+	 * limited to 79 characters plus the null-termination character.
+	 */
+	DEV_CHAN_POS_NMEA,
+
+	/**
+	 * An Euler angle, which consists of the three familiar X/Y/Z
+	 * headings.
+	 *
+	 * Expressed as a 3-vector (XYZ) of degrees.
+	 */
+	DEV_CHAN_ORIENT_EULER,
+
+	/**
+	 * A quaternion, which is generally used to represent a value in
+	 * three dimensional space, avoiding the problem of gimbal lock
+	 * associated with 3-vector Euler angles.
+	 *
+	 * Quaternions consist of four elements: a 3-vector and a scalar.
+	 */
+	DEV_CHAN_ORIENT_QUAT,
+
+	/**
+	 * An indication of tilt in degrees on a single axis relative to an
+	 * unspecified reference plane.
+	 *
+	 * Expressed in degrees (+/-180).
+	 */
+	DEV_CHAN_ORIENT_TILT,
+
+
+	/**
+	 * Pressure, which is a measurement of the force applied
+	 * perpendicularly to a specific surface area.
+	 *
+	 * Expressed in hectopascal (hPa)
+	 */
+	DEV_CHANNEL_PRESS,
+
+	/**
+	 * Barometric (or 'atmospheric') pressure, which is a measurement of
+	 * the pressure within the atmosphere. Related to the concept of
+	 * 'hydrostatic pressure' caused by the weight of the air above the
+	 * point of measure.
+	 *
+	 * TODO: Would millibar (mb) be a better unit choice, or should this
+	 * remain consist with the more general pressure channel unit (hPa)?
+	 *
+	 * Expressed in hectopascal (hPa)
+	 */
+	DEV_CHANNEL_PRESS_BAROM,
+
+	/**
+	 * Particulate matter.
+	 *
+	 * Expressed as micrograms per cubic meter (ug/m^3).
+	 */
+	DEV_CHAN_PART_MAT,
+
+	/**
+	 * Generic parts per million measurement. The exact type of matter
+	 * being measured should be indicated with an appropriate channel
+	 * attribute.
+	 *
+	 * Expressed as parts per million (ppm).
+	 */
+	DEV_CHAN_PPM,
+
+	/**
+	 * Detected CO2 levels.
+	 *
+	 * Expressed as parts per million (ppm).
+	 */
+	DEV_CHAN_PPM_CO2,
+
+	/**
+	 * Detected volatile organic compound (voc) levels.
+	 *
+	 * Expressed as parts per million (ppm).
+	 */
+	DEV_CHAN_PPM_VOC,
+
+	/**
+	 * User-defined channel. This can be used at the application level for
+	 * whetever purpose the driver requires, provided it does not overlap
+	 * with one of the existing standard channel definitions.
+	 *
+	 * If multiple user defined channels are present, they should be
+	 * clearly distinguished by use of appropriate channel attributes.
+	 *
+	 * Expressed as a user-defined unit and scale. The exact unit type and
+	 * scale should be clearly documented in the driver making use of this
+	 * channel type.
+	 */
+	DEV_CHAN_USER_DEF = ZIO_CHAN_LIMIT,
+} zio_dev_chan_type_t;
```

Comment:
STEP_COUNT, STEP_DETECT, TAP, DOUBLE_TAP, WAKEUP, GESTURE?

Maybe we should handle also this types.

Reply:
Thanks for the feedback ... slow to reply because I've been thinking about this, sorry. There's a point where it becomes blurry what is a **'channel'* or an **'attribute'**, and what is better served by something like an **'event'**, which is seems like `STEP_COUNT`, `STEP_DETECT`, `TAP`, etc., would be. 

That also gets me thinking about the concept of **attributes with state**, though, which are probably better described as **`properties`** at the sensor API level, like `STEPS_DETECTED` which would be tracked over time with an option to reset.

I don't want to over complicate the sensor API with extra options just to satisfy my need to have a box for everything, but at what point does something like 'GESTUR' become specific to a single driver's API or user code, and when should we have generic solution at the sensor API level (gesture is hard for that if you want to define gesture types)?

The idea of state management is a good question to consider, though, and how to handle that with the current channel + attribute approach, and we may need or benefit from a concept like `property` as well?

Runtime contributed a [stats module](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/stats/stats.c) that I've used numerous times to track incrementing variables like this (it works with mcumgr from the command-line which is nice!), and that might be a viable solution using an existing API ... but since that subsystem is more intended for device diagnostics and debugging in the field, it's may be forcing it to do something it isn't designed for and we should have this directly in the sensor API.

Thanks for getting the wheels turning on the events and properties issues, though, and any thoughts are highly welcome?